documentation/tuning_guide.markdown

```diff
@@ -0,0 +1,156 @@
+a network connection as part of the processing of a report, and if there is a chance
+that the network operation is slow and will block on I/O for some period of time,
+then it may make sense to have more JRuby instances than the number of cores,
+because the JVM will be smart enough to suspend the thread that is handling those
+kinds of requests and use the CPUs for other work, assuming there are still JRuby
+instances available in the pool.  In a case like this you might want to set
+`max-active-instances` to a value higher than the number of CPUs.)
+
+At this point you may be wondering "what's the downside to just setting
+`max-active-instances` to a really high value?"  The answer to this question, in
+a nutshell, is "memory usage".  This brings us to the other extremely important
+Puppet Server.
+
+### JVM Heap Size
+
+The JVM's "max heap size" controls the maximum amount of (heap*[[1]](#footnotes)
+memory that the JVM process is allowed to request from the operating system.  You
+can set this value via the `-Xmx` command-line argument at JVM startup.  (In the
+case of Puppet Server, you'll find this setting in the "defaults" file for Puppet
+Server for your operating system; this will generally be something like
+`/etc/sysconfig/puppetserver` or `/etc/defaults/puppetserver`.)
+
+If your application's memory usage approaches this value, the JVM will try to
+get more aggressive with garbage collection to free up memory.  (In certain
+situations you may see increased CPU activity related to this garbage collection.)
+If the JVM is unable to recover enough memory to keep the application running
+smoothly, you will eventually encounter an `OutOfMemoryError`, and the process
+will shut down.  (For Puppet Server, we also use a JVM argument called
+`-XX:HeapDumpOnOutOfMemoryError` to cause the JVM to dump an `.hprof` file to
+disk, which is basically a memory snapshot from the point in time where the
+error occurred; this can be loaded into various profiling tools to get a better
+understanding of where the memory was being used.)
+
+(Note that there is another setting, "min heap size", that is controlled via
+the -Xms setting; [Oracle recommends](http://www.oracle.com/technetwork/java/gc-tuning-5-138395.html#0.0.0.%20Total%20Heap|outline)
+setting this value to the same value that you use for -Xmx.)
+
+The most important factor when determining the max heap size for Puppet Server
+is the value of `max-active-instances`.  Each JRuby instance needs to load up
+a copy of the Puppet Ruby code, and then needs some amount of memory overhead
+for all of the garbage that gets generated during a Puppet catalog compilation.
+Also, the memory requirements will vary based on how many Puppet modules you
+have in your module path, how much hiera data you have, etc.  At this time we're
+estimating that a reasonable ballpark figure is about 512MB of RAM per JRuby
+instance, give or take a bit based on the size of your Puppet codebase.
```

Comment:
I might add to or change the wording in this paragraph to more strongly indicate that memory usage per JRuby can vary wildly and, while 512MB may be a reasonable starting spot for most installations, if a user cares about tuning that parameter they can expect to end up with something very different.